sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/CoreDataflowNodeProcessor.scala

```diff
@@ -0,0 +1,229 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pipelines.graph
+
+import java.util.concurrent.{ConcurrentHashMap, ConcurrentLinkedQueue}
+
+import scala.jdk.CollectionConverters._
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.pipelines.graph.DataflowGraphTransformer.{
+  TransformNodeFailedException,
+  TransformNodeRetryableException
+}
+
+/**
+ * Processor that is responsible for analyzing each flow and sort the nodes in
+ * topological order
+ */
+class CoreDataflowNodeProcessor(rawGraph: DataflowGraph) {
+
+  private val flowResolver = new FlowResolver(rawGraph)
+
+  // Map of input identifier to resolved [[Input]].
+  private val resolvedInputs = new ConcurrentHashMap[TableIdentifier, Input]()
+  // Map & queue of resolved flows identifiers
+  // queue is there to track the topological order while map is used to store the id -> flow
+  // mapping
+  private val resolvedFlowNodesMap = new ConcurrentHashMap[TableIdentifier, ResolvedFlow]()
+  private val resolvedFlowNodesQueue = new ConcurrentLinkedQueue[ResolvedFlow]()
+
+  private def processUnresolvedFlow(flow: UnresolvedFlow): ResolvedFlow = {
+    val resolvedFlow = flowResolver.attemptResolveFlow(
+      flow,
+      rawGraph.inputIdentifiers,
+      resolvedInputs.asScala.toMap
```

Comment:
shall we let `attemptResolveFlow` to take java hash map so that we can save this scala -> java conversion?